primitives/block/src/equivocation_proof.rs

```diff
@@ -0,0 +1,455 @@
+    /// Address of the offending validator.
+    validator_address: Address,
+    /// Hash of proposal number 1.
+    proposal_hash1: Option<Blake2sHash>,
+    /// Hash of proposal number 2.
+    proposal_hash2: Option<Blake2sHash>,
+    /// Aggregate signature for proposal 1.
+    signature1: AggregateSignature,
+    /// Aggregate signature for proposal 2.
+    signature2: AggregateSignature,
+    /// Signers for proposal 1.
+    signers1: BitSet,
+    /// Signers for proposal 2.
+    signers2: BitSet,
+}
+
+impl DoubleVoteProof {
+    /// The maximum size of a double proposal proof.
+    pub const MAX_SIZE: usize = 2 * MacroHeader::MAX_SIZE
+        + 2 * nimiq_serde::option_max_size(Blake2sHash::SIZE)
+        + 2 * AggregateSignature::SIZE
+        + 2 * BitSet::max_size(Policy::SLOTS as usize);
+
+    pub fn new(
+        tendermint_id: TendermintIdentifier,
+        validator_address: Address,
+        mut proposal_hash1: Option<Blake2sHash>,
+        mut proposal_hash2: Option<Blake2sHash>,
+        mut signature1: AggregateSignature,
+        mut signature2: AggregateSignature,
+        mut signers1: BitSet,
+        mut signers2: BitSet,
+    ) -> DoubleVoteProof {
+        if proposal_hash1 > proposal_hash2 {
+            mem::swap(&mut proposal_hash1, &mut proposal_hash2);
+            mem::swap(&mut signature1, &mut signature2);
+            mem::swap(&mut signers1, &mut signers2);
+        }
+        DoubleVoteProof {
+            tendermint_id,
+            validator_address,
+            proposal_hash1,
+            proposal_hash2,
+            signature1,
+            signature2,
+            signers1,
+            signers2,
+        }
+    }
+
+    /// Block number.
+    pub fn block_number(&self) -> u32 {
+        self.tendermint_id.block_number
+    }
+    /// Address of the offending validator.
+    pub fn validator_address(&self) -> &Address {
+        &self.validator_address
+    }
+
+    /// Verify the validity of a double proposal proof.
+    pub fn verify(&self, validators: &Validators) -> Result<(), EquivocationProofError> {
+        // Check that the proposals are not equal and in the right order:
+        match self.proposal_hash1.cmp(&self.proposal_hash2) {
+            Ordering::Less => {}
+            Ordering::Equal => return Err(EquivocationProofError::SameHeader),
+            Ordering::Greater => return Err(EquivocationProofError::WrongOrder),
+        }
+
+        let validator = match validators.get_validator_by_address(&self.validator_address) {
+            None => return Err(EquivocationProofError::InvalidValidatorAddress),
+            Some(v) => v,
+        };
+
+        // Check that at least one of the validator's slots is actually contained in both signer sets.
+        if !validator
+            .slots
+            .clone()
+            .any(|s| self.signers1.contains(s as usize) && self.signers2.contains(s as usize))
+        {
+            return Err(EquivocationProofError::SlotMismatch);
```

Comment:
Could be more specific like `EquivocationProofError::NoOverlap`.

Reply:
Done.